### ISSUE TITLE
Images used by textures with different color spaces? 

### ISSUE THREAD

**Issue by donmccurdy** [KhronosGroup/glTF#2308]
The spec discusses image color spaces under [_Images, Section 3.8.3_](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#images):

> Effective transfer function (encoding) is defined by a glTF object that refers to the image (in most cases it’s a texture that is used by a material).

I believe this implies that the same _texture_ object cannot be associated with multiple color spaces, and so cannot be used both as a `baseColorTexture` and `occlusionTexture`. Would it be valid for the same _image_ object to be used by two textures in those material slots?

I've run into a question in glTF Transform, where it merges two identical textures, resulting in a color space difference in three.js (which expects 1 color space per glTF texture), and I'm not sure  which is correct.

- https://github.com/donmccurdy/glTF-Transform/discussions/1061

**Comment by bghgary** [comment]
> Would it be valid for the same image object to be used by two textures in those material slots?

Doing this is probably unusual, but I think it is valid. @lexaknyazev What do you think? For reference, if you load the two models in the referenced glTF-Transform issue in Babylon.js sandbox, they look the same before and after.

![image](https://github.com/KhronosGroup/glTF/assets/19158740/d04cfa84-6807-4bf0-adcd-410b907833f2)

**Comment by julienduroure** [comment]
Hello,
I got this case in a report last year: 
https://github.com/KhronosGroup/glTF-Blender-IO/issues/1800

**Comment by Samsy** [comment]
> > Would it be valid for the same image object to be used by two textures in those material slots?
> 
> Doing this is probably unusual, but I think it is valid. @lexaknyazev What do you think? For reference, if you load the two models in the referenced glTF-Transform issue in Babylon.js sandbox, they look the same before and after.
> 
> ![image](https://user-images.githubusercontent.com/19158740/260588549-d04cfa84-6807-4bf0-adcd-410b907833f2.png)

I just checked Babylon viewer and indeed they look the same, 
I inspected with spector, babylon viewer isn't using sRGB color spacing which was the problem originally, this may be why they both look the same

**Comment by bghgary** [comment]
> babylon viewer isn't using sRGB color spacing which was the problem originally

I think you mean the sandbox doesn't use EXT_srgb formats. If so, yes, by default, we have this turned off because of [a Chromium issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1256340). If you use https://3dcommerce.babylonjs.com, it will use EXT_srgb formats. The result is the same though.

**Comment by donmccurdy** [comment]
FWIW, three.js has moved to use `texStorage2D` which avoids the Chromium performance issues.  We do still see similar performance issues on video textures, and use inline sRGB decoding in the shader to handle that instead.

While investigating support for wide gamut colorspaces in WebGL and three.js, I've encountered an issue that might be relevant here. Images are unpacked by WebGL into a target color space identified by [`gl.unpackColorSpace`](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/unpackColorSpace). The web browser infers the source color space of an image from its ICC profile, and uses that to make the required conversion into the unpack color space. While the glTF Spec advises setting the "UNPACK_COLORSPACE_CONVERSION_WEBGL flag to NONE," we can no longer do that in scenes whose textures use more than one set of color primaries, without re-implementing the entire image decoding pipeline ourselves (and we aren't going to do that).

As a result — I think the glTF specification's current advice about ignoring ICC Profiles is still reasonable, however, I *do not* believe we should be encouraging situations where the embedded ICC Profile cannot possibly be accurate, as would be the case when the image contains mixed data from different color spaces.

**Comment by lexaknyazev** [comment]
> we can no longer do that in scenes whose textures use more than one set of color primaries

By definition, glTF 2.0 textures use only ITU-R BT.709 primaries. Detecting and rejecting assets that have something else may be very difficult because differently encoded ICC profiles may define the same color space (e.g., precomputed LUTs vs parameter-based built-in functions).

**Comment by donmccurdy** [comment]
@lexaknyazev What I mean is that the textures embedded in a glTF 2.0 asset might not be the only textures that a client implementation is loading. We might have textures from other sources, properly identified in some other color space. In that situation, we cannot simply disable `gl.unpackColorSpace` as the spec suggests, we need it.

The case that I think we might want to detect and reject would be, for example, if a glTF 2.0 asset contains a texture that is used both for `baseColorTexture` and for `occlusionTexture`. A client with `gl.unpackColorSpace` enabled cannot load this asset correctly.

But even then, this is probably a minefield. How should a `normalTexture` be configured to ensure it doesn't ever go through a color space conversion? I don't think there's an ICC Profile corresponding to "non color data"... 😕

**Comment by lexaknyazev** [comment]
> What I mean is that the textures embedded in a glTF 2.0 asset might not be the only textures that a client implementation is loading.

This is clearly out of scope of the glTF spec.

> If a glTF 2.0 asset contains a texture that is used both for `baseColorTexture` and for `occlusionTexture`.

Pixel values from non-color textures, e.g., occlusion, normal, roughness, etc, must be used as-is, i.e., ignoring any color primaries or transfer function information. So they must be uploaded with all color space related conversions disabled.

Pixel values stored in color textures referred from a glTF 2.0 asset **must** use BT.709 color primaries and sRGB transfer function. If the current display uses, e.g., Display-P3 primaries, it's an engine's responsibility to ensure correct look.

> a texture that is used both for `baseColorTexture` and for `occlusionTexture`

This is not explicitly disallowed (although could be a validation warning). An engine may have to upload the texture twice in this case using different options since both `unpackColorSpace` and `UNPACK_COLORSPACE_CONVERSION_WEBGL` are mutable.

**Comment by donmccurdy** [comment]
I feel that a validation warning for “textures used both for baseColorTexture and for occlusionTexture” would be a good thing. I don't think any other action, or changes to the spec, are being requested in this thread.

**Comment by UX3D-haertl** [comment]
> This is not explicitly disallowed (although could be a validation warning). An engine may have to upload the texture twice in this case using different options since both unpackColorSpace and UNPACK_COLORSPACE_CONVERSION_WEBGL are mutable.

Currently, sample viewer does not handle this case. I would have assumed that this is not intended and glTF textures map to WebGL texture 1 to 1. But I guess the spec technically allows using the same texture in mulitple material properties which expect different color spaces. In that case we should probably implement this feature in sample viewer even if it is a fringe case.

**Comment by oss-haertl** [comment]
I had another look at this issue, and it seems like KHR_texture_basisu prohibits the use of a KTX texture for different color spaces.

> - Color space information in the DFD MUST match the expected usage, namely:
>   - For textures with **color data** (e.g., base color maps),
>     - `colorPrimaries` MUST be `KHR_DF_PRIMARIES_BT709`;
>     - `transferFunction` MUST be `KHR_DF_TRANSFER_SRGB`.
>   - For textures with **non-color data** (e.g., normal maps),
>     - `colorPrimaries` MUST be `KHR_DF_PRIMARIES_UNSPECIFIED`;
>     - `transferFunction` MUST be `KHR_DF_TRANSFER_LINEAR`.

It obviously makes sense for KTX, but it feels somehow inconsistent that other image formats behave differently.

**Comment by donmccurdy** [comment]
In my opinion the specified behavior of traditional image formats in glTF (clients MUST ignore color space metadata in PNG, JPG, ...) is a compromise: ideally they would also also contain correct metadata. But practically PNG or JPG metadata is going to be wrong so often that it's almost worthless. I'm not sure it's even possible to annotate a PNG or JPG image with metadata indicating that it's "non-color data", as would be needed for metal/rough/occlusion textures. But because KTX2 is new and the tools are more modern, getting the metadata right was not a lost cause yet, it seemed worthwhile to try, and (so far?) that seems to be successful.

**Comment by lexaknyazev** [comment]
> the specified behavior of traditional image formats in glTF (clients MUST ignore color space metadata in PNG, JPG, ...) is a compromise

Yes, exactly. There's no easy way to fully know the color profile of an arbitrary image stored in these formats as it's not feasible to ask glTF implementations to support ICC profile parsing and decoding. FWIW, even the [WebCodecs API](https://w3c.github.io/webcodecs/) won't help much with this.

So the only reliable option is to always ignore embedded ICC data and treat those image files as either using BT.709 primaries with sRGB non-linear encoding or as containing non-color values (like roughness) with linear encoding.
